src/esbuild/index.ts

```diff
@@ -0,0 +1,64 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the MIT License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2019-Present Datadog, Inc.
+
+/* eslint-disable no-console */
+
+import { PluginBuild, BuildResult } from 'esbuild';
+
+import { wrapPlugins, getResults as getPluginsResults } from './plugins';
+import { BaseClass } from '../BaseClass';
+import { Options } from '../types';
+import { getModulesResults } from './modules';
+
+class BuildPluginClass extends BaseClass {
+    constructor(opts: Options) {
+        super(opts);
+        // This options is filled by webpack in the webpack plugin.
+        this.options.context = process.cwd();
+    }
+    setup(build: PluginBuild) {
+        const startBuild = Date.now();
+        // We force esbuild to produce its metafile.
+        build.initialOptions.metafile = true;
+        wrapPlugins(this, build, build.initialOptions.plugins);
+        build.onEnd(async (result: BuildResult) => {
+            const { plugins, modules } = getPluginsResults();
+            // We know it exists since we're setting the option earlier.
+            const metaFile = result.metafile!;
+            const moduleResults = getModulesResults(metaFile, this.options.context);
+
+            this.addContext({
+                start: startBuild,
+                report: {
+                    timings: {
+                        tapables: plugins,
+                        modules,
+                    },
+                    dependencies: moduleResults,
+                },
+                bundler: {
+                    esbuild: {
```

Comment:
Does it mean the metrics will be uploaded in separate "namespaces"? How will that interact with things like the budget plan?

Reply:
Nope, this doesn't change the metric's namespace.
The metric's namespace is decided at the integration level, in the settings.
Right now, we have it setup as `dd.web_ui.webpack` for our webpack build.

https://github.com/DataDog/web-ui/blob/8d8e8bed1a0c4dcca1a9da42846c94390e0b91a5/webpack.config.js#L513

Here, it's just for better handling esbuild's specific data internally.

That being said, I think we should use a different namespace for esbuild.
It will ease the migration for the budget plan too.
Otherwise, during the migration, we may end up with metrics overwriting each others (between webpack and esbuild builds) and the budget will go awry.